virtcontainers/clh.go

```diff
@@ -877,461 +864,258 @@ func MaxClhVCPUs() uint32 {
+	cscAPIsocket string = "--api-socket"
 	cscLogFile   string = "--log-file"
-	cscMemory    string = "--memory"
-	cscNet       string = "--net"
-	cscRng       string = "--rng"
-	cscSerial    string = "--serial"
-	cscVsock     string = "--vsock"
 )
 
-type CommandLineBuilder interface {
-	AddKernelParameters(cmdline []Param)
-	SetConsole(console *CLIConsole)
-	SetCpus(cpus *CLICpus)
-	SetDisk(disk *CLIDisk)
-	SetFs(fs *CLIFs)
-	SetKernel(kernel *CLIKernel)
-	SetMemory(memory *CLIMemory)
-	AddNet(net CLINet)
-	SetRng(rng *CLIRng)
-	SetSerial(serial *CLISerialConsole)
-	SetVsock(vsock *CLIVsock)
-	SetAPISocket(apiSocket *CLIAPISocket)
-	SetLogFile(logFile *CLILogFile)
-	GetCommandLine() (*CommandLine, error)
-}
-
-type CLIOption interface {
-	Build(cmdline *CommandLine)
-}
-
-type CommandLine struct {
-	args []string
-}
-
-//**********************************
-// The (virtio) Console
-//**********************************
-type CLIConsole struct {
-	consoleType string
-	filePath    string
-	iommu       bool
-}
+//****************************************
+// The kernel command line
+//****************************************
 
-func (o *CLIConsole) Build(cmdline *CommandLine) {
+func kernelParamsToString(params []Param) string {
 
-	cmdline.args = append(cmdline.args, cscConsole)
+	var paramBuilder strings.Builder
+	for _, p := range params {
+		paramBuilder.WriteString(p.Key)
+		if len(p.Value) > 0 {
 
-	consoleArg := ""
-	if o.consoleType == cctFILE {
-		consoleArg = o.consoleType + "=" + o.filePath
-		if o.iommu {
-			consoleArg += ",iommu=on"
-		} else {
-			consoleArg += ",iommu=off"
+			paramBuilder.WriteString("=")
+			paramBuilder.WriteString(p.Value)
 		}
-	} else {
-		consoleArg = o.consoleType
+		paramBuilder.WriteString(" ")
 	}
-
-	cmdline.args = append(cmdline.args, consoleArg)
+	return strings.TrimSpace(paramBuilder.String())
 }
 
-//**********************************
-// The serial port
-//**********************************
-type CLISerialConsole struct {
-	consoleType string
-	filePath    string
-}
+//****************************************
+// API calls
+//****************************************
+func (clh *cloudHypervisor) isClhRunning(timeout uint) (bool, error) {
 
-func (o *CLISerialConsole) Build(cmdline *CommandLine) {
+	pid := clh.state.PID
 
-	cmdline.args = append(cmdline.args, cscSerial)
-	if o.consoleType == cctFILE {
-		cmdline.args = append(cmdline.args, o.consoleType+"="+o.filePath)
-	} else {
-		cmdline.args = append(cmdline.args, o.consoleType)
+	// Check if clh process is running, in case it is not, let's
+	// return from here.
+	if err := syscall.Kill(pid, syscall.Signal(0)); err != nil {
+		return false, nil
 	}
 
-}
-
-//**********************************
-// The API socket
-//**********************************
-type CLIAPISocket struct {
-	socketPath string
-}
+	timeStart := time.Now()
+	cl := clh.client()
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), clhAPITimeout*time.Second)
```

Comment:
If, at the end of iteration N, the remaining time (w.r.t. timeout) is shorter than clhAPITimeout, we would still wait for clhAPITimeout in iteration N+1. This would make the total wait longer than timeout.

Is this Okay ?